src/Hazelcast.Net.Tests/Networking/RegressionWithRealNetworkTest.cs

```diff
@@ -0,0 +1,249 @@
+            await MembersOneTimeTearDown();
+        }
+
+        [Test]
+        [TestCase(true)]
+        [TestCase(false)]
+        public async Task TestClientConnectionBeforeServerReady(bool smartRouting)
+        {
+            using var _ = HConsoleForTest();
+
+            var clientTask = CreateAndStartClientAsync(options =>
+            {
+                options.Networking.SmartRouting = smartRouting;
+                options.Networking.ConnectionRetry.ClusterConnectionTimeoutMilliseconds = int.MaxValue;
+                options.Networking.ReconnectMode = ReconnectMode.ReconnectAsync;
+                options.ClusterName = RcCluster.Id;
+            });
+
+            //Let the client tries to connect before cluster is ready
+            await Task.Delay(500);
+
+            var member = await AddMember();
+
+            var client = await clientTask;
+
+            await AssertEx.SucceedsEventually(() =>
+            {
+                Assert.AreEqual(ClientState.Connected, client.State);
+            }, 10_000, 500);
+        }
+
+        [Test]
+        [Timeout(20_000)]
+        [TestCase(true, "localhost", "127.0.0.1")]
+        [TestCase(true, "127.0.0.1", "localhost")]
+        [TestCase(true, "localhost", "localhost")]
+        [TestCase(false, "localhost", "127.0.0.1")]
+        [TestCase(false, "127.0.0.1", "localhost")]
+        [TestCase(false, "localhost", "localhost")]
+        public async Task TestConnectionCountAfterClientReconnect(bool smartRouting, string memberAddress, string clientAddress)
+        {
+            using var _ = HConsoleForTest();
+            #region SetUp
+            var memberConfig = GetMemberConfigWithAddress(memberAddress);
+            var customCluster = await RcClient.CreateClusterAsync(memberConfig);
+            var member = await RcClient.StartMemberAsync(customCluster.Id);
+            var client = await CreateAndStartClientAsync(options =>
+           {
+               options.Networking.SmartRouting = smartRouting;
+               options.Networking.ConnectionRetry.ClusterConnectionTimeoutMilliseconds = 10_000;
+               options.Networking.ReconnectMode = ReconnectMode.ReconnectAsync;
+               options.Networking.Addresses.Clear();
+               options.Networking.Addresses.Add(clientAddress + ":" + member.Port);
+               options.ClusterName = customCluster.Id;
+           });
+            #endregion
+
+            var map = await client.GetMapAsync<int, int>("TestMap_" + DateTime.UtcNow.Millisecond);
+            var clientImp = (HazelcastClient)client;
+
+            Assert.AreEqual(1, clientImp.Cluster.Connections.Count);
+            await map.PutAsync(1, 1);
+
+            await RcClient.ShutdownMemberAsync(customCluster.Id, member.Uuid);
+
+            await AssertEx.SucceedsEventually(() =>
+            {
+                Assert.AreEqual(ClientState.Disconnected, client.State);
+            }, 10_000, 500);
+
```

Comment:
Add a test for connections count, should be zero?

Reply:
Fixed at 1994fec7c1ec5775d246230ddb9ab7f388640b4b